### PR TITLE
refactor _d_arraysetlengthT()

### PR DESCRIPTION
**Pull request by WalterBright**
Implement:

1. eliminate gotos
2. minimize scope of variables
3. simplify control flow
4. use const
5. reduce indentation levels

More can be done, but this is enough for the moment.

### CODE REVIEW

**Comment by dejlek** [comment]
It looks OK to my noobish eyes. :) I would however try to break the function into into few smaller pieces (functions), if possible, to make it easier to reason about...

File: src/rt/lifetime.d [hunk 1]
```diff
@@ -1456,172 +1457,179 @@ do
...
+                        info.size = u;
+                        if(__setArrayAllocLength(info, newsize + offset, isshared, tinext, size + offset))
                         {
-                            // add this to the cache, it wasn't present previously.
-                            __insertBlkInfoCache(info, null);
+                            if(!isshared)
+                                __insertBlkInfoCache(info, bic);
+                            memset(newdata + size, 0, newsize - size);
+                            *p = newdata[0 .. newlength];
+                            return *p;
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply src/rt/lifetime.d]: There were a number of lines not tested before.

**Comment by WalterBright** [comment]
> and needs similar treatment.

Yes.

**Comment by schveiguy** [comment]
@wilzbach why isn't this merging?

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
> why isn't this merging?

Because Jenkins didn't reply. That's sadly a known issue with Jenkins. Though it happens rarely and closing and reopening triggers a new Jenkins build.

**Pull request merged**